accounts/abi/bind/bind.go

```diff
@@ -280,36 +323,328 @@ func Bind(types []string, abis []string, bytecodes []string, fsigs []map[string]
+			}
+			normalizedArguments[i].Name = fmt.Sprintf("%s%d", normalizedArguments[i].Name, index)
+		}
+		if hasStruct(input.Type) {
+			cb.binder.BindStructType(input.Type)
+		}
+	}
+	return normalizedArguments
+}
+
+// bindEvent normalizes an event and registers it to be emitted in the bindings.
+func (cb *contractBinder) bindEvent(original abi.Event) error {
+	// Skip anonymous events as they don't support explicit filtering
+	if original.Anonymous {
+		return nil
 	}
-	// For all others just return as is for now
-	return buffer.String(), nil
+	normalizedName, err := cb.RegisterEventIdentifier(original.Name)
+	if err != nil {
+		return err
+	}
+
+	normalized := original
+	normalized.Name = normalizedName
+	normalized.Inputs = cb.normalizeErrorOrEventFields(original.Inputs)
+	cb.events[original.Name] = &tmplEvent{Original: original, Normalized: normalized}
+	return nil
+}
+
+// bindEvent normalizes an error and registers it to be emitted in the bindings.
+func (cb *contractBinder) bindError(original abi.Error) error {
+	normalizedName, err := cb.RegisterErrorIdentifier(original.Name)
+	if err != nil {
+		return err
+	}
+
+	normalized := original
+	normalized.Name = normalizedName
+	normalized.Inputs = cb.normalizeErrorOrEventFields(original.Inputs)
+	cb.errors[original.Name] = &tmplError{Original: original, Normalized: normalized}
+	return nil
 }
 
-// bindType is a set of type binders that convert Solidity types to some supported
-// programming language types.
-var bindType = map[Lang]func(kind abi.Type, structs map[string]*tmplStruct) string{
-	LangGo: bindTypeGo,
+// BindV2 generates a Go wrapper around a contract ABI. This wrapper isn't meant
+// to be used as is in client code, but rather as an intermediate struct which
+// enforces compile time type safety and naming convention as opposed to having to
+// manually maintain hard coded strings that break on runtime.
+func BindV2(types []string, abis []string, bytecodes []string, pkg string, libs map[string]string, aliases map[string]string) (string, error) {
+	b := binder{
+		contracts: make(map[string]*tmplContractV2),
+		structs:   make(map[string]*tmplStruct),
+		aliases:   aliases,
+	}
+	for i := 0; i < len(types); i++ {
+		// Parse the actual ABI to generate the binding for
+		evmABI, err := abi.JSON(strings.NewReader(abis[i]))
+		if err != nil {
+			return "", err
+		}
+		// Strip any whitespace from the JSON ABI
+		strippedABI := strings.Map(func(r rune) rune {
+			if unicode.IsSpace(r) {
+				return -1
+			}
+			return r
+		}, abis[i])
+
+		for _, input := range evmABI.Constructor.Inputs {
+			if hasStruct(input.Type) {
+				bindStructType(input.Type, b.structs)
+			}
+		}
+
+		cb := contractBinder{
+			binder: &b,
+			calls:  make(map[string]*tmplMethod),
+			events: make(map[string]*tmplEvent),
+			errors: make(map[string]*tmplError),
+
+			callIdentifiers:  make(map[string]bool),
+			errorIdentifiers: make(map[string]bool),
+			eventIdentifiers: make(map[string]bool),
+		}
```

Comment:
```suggestion
		cb := newContractBinder(b)
```
Assuming `b` is a pointer